src/realm/collection.cpp

```diff
@@ -93,46 +93,77 @@ void check_for_last_unresolved(BPlusTree<ObjKey>* tree)
 
 Collection::~Collection() {}
 
-Mixed Collection::get_any(Mixed val, Path::const_iterator begin, Path::const_iterator end, Allocator& alloc)
+void Collection::get_any(QueryCtrlBlock& ctrl, Mixed val, size_t index)
 {
-    auto path_size = end - begin;
-    if (val.is_type(type_Dictionary) && begin->is_key()) {
-        Array top(alloc);
-        top.init_from_ref(val.get_ref());
-
-        BPlusTree<StringData> keys(alloc);
+    auto path_size = ctrl.path.size() - index;
+    PathElement& pe = ctrl.path[index];
+    if (val.is_type(type_Dictionary) && (pe.is_key() || pe.is_all())) {
+        auto ref = val.get_ref();
+        if (!ref)
+            return;
+        Array top(ctrl.alloc);
+        top.init_from_ref(ref);
+
+        BPlusTree<StringData> keys(ctrl.alloc);
         keys.set_parent(&top, 0);
         keys.init_from_parent();
-        size_t ndx = keys.find_first(StringData(begin->get_key()));
-        if (ndx != realm::not_found) {
-            BPlusTree<Mixed> values(alloc);
+        size_t start = 0;
+        if (size_t finish = keys.size()) {
+            if (pe.is_key()) {
+                start = keys.find_first(StringData(pe.get_key()));
+                if (start == realm::not_found) {
+                    if (pe.get_key() == "@keys") {
```

Comment:
Not a blocker, but I'd almost expect something like `pe.is_keys()` here rather than checking hard coded strings this deep in the code.